### DISCUSSION TITLE
Skip clang tools on ExternalProject

### DISCUSSION THREAD

**Discussion by ChristopherPisz** [friendlyanon/cmake-init#62] [Q&A]
I bring in gtest as an external project. How can I get the build to skip running clang tools on the external project, but still run them on my source code?

I've got a gtest.cmake in project/externals that looks like:

```cmake
set(GTEST_PREFIX "${CMAKE_BINARY_DIR}/externals/gtest")
set(GTEST_INCLUDE_DIR "${GTEST_PREFIX}/src/googletest/googletest/include")
set(GTEST_LIB_DIR "${GTEST_PREFIX}/src/googletest-build/lib")

ExternalProject_Add( 
        googletest
        PREFIX "${GTEST_PREFIX}"
        GIT_REPOSITORY https://github.com/google/googletest
        GIT_TAG main
        INSTALL_COMMAND ""
        UPDATE_COMMAND ""
)

include_directories(${GTEST_INCLUDE_DIR})
link_directories(${GTEST_LIB_DIR})
```

and it is used in my tests via the following cmake entries in the test/CmakeLists.txt

```cmake
target_link_libraries(gtest-myproject PRIVATE
        myproject::myproject
        Threads::Threads
        gtest
        gmock
        )
add_dependencies(gtest-myproject
        googletest)
gtest_discover_tests(gtest-myproject)
````

**Comment by friendlyanon** [comment]
ExternalProject is more useful for orchestrating superbuilds. cmake-init allows you to use a package manager, just use one of them to bring in dependencies.

**Reply by ChristopherPisz** [reply]
I dunno what that is or how to go about it. :(
Google says bunches of people made their own projects that serve as package managers on their individual githubs. I need something that is popular or already supported enough to convince devops to have it running on a build machine. Is anything already built in? If not, is there a widely supported selection?

**Reply by friendlyanon** [reply]
Conan and vcpkg are options that are provided by cmake-init. One is from JFrog, the other from Microsoft.

**Reply by ChristopherPisz** [reply]
Tried using CPM.cmake and still get all the warning about googletest includes.
I am looking over the conan docs and it's going to be too much to convince my company to install yet another piece of software via python and do all the cmake edits it requires.

**Reply by friendlyanon** [reply]
CPM is just a wrpper around FetchContent which is a wrapper around ExternalProject. All forms of vendoring that must be avoided.

Conan doesn't require any changes to CMake code to use. Please just use the package manager option from cmake-init and experiment with the output.

**Reply by ChristopherPisz** [reply]
Tried generating a test project with the conan option after installing all optional software listed in the readme
I get the following result when opening the project in CLion:

`/snap/clion/193/bin/cmake/linux/bin/cmake -DCMAKE_BUILD_TYPE=Debug --preset dev -S /home/cpisz/git-workspace/testproject -B /home/cpisz/git-workspace/testproject/build/dev
Preset CMake variables:

  BUILD_MCSS_DOCS="ON"
  CMAKE_CXX_CLANG_TIDY="clang-tidy;--header-filter=/home/cpisz/git-workspace/testproject/*"
  CMAKE_CXX_CPPCHECK="cppcheck;--inline-suppr"
  CMAKE_CXX_EXTENSIONS="OFF"
  CMAKE_CXX_FLAGS="-Wall -Wextra -Wpedantic -Wconversion -Wsign-conversion -Wcast-qual -Wshadow -Wformat=2 -Wundef -Werror=float-equal"
  CMAKE_CXX_STANDARD="17"
  CMAKE_CXX_STANDARD_REQUIRED="ON"
  CMAKE_POLICY_DEFAULT_CMP0091="NEW"
  CMAKE_TOOLCHAIN_FILE="/home/cpisz/git-workspace/testproject/conan/conan_toolchain.cmake"
  testproject_DEVELOPER_MODE="ON"

CMake Error at /snap/clion/193/bin/cmake/linux/share/cmake-3.22/Modules/CMakeDetermineSystem.cmake:130 (message):
  Could not find toolchain file:
  /home/cpisz/git-workspace/testproject/conan/conan_toolchain.cmake
Call Stack (most recent call first):
  CMakeLists.txt:5 (project)


CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!

[Failed to reload]
`
It looks like cmake-init wants to use fmt as an example, while the getting started on the conan page wants to use a txt file and poco as an example, so I am not sure if I am supposed to do something obvious, or if something went wrong.

**Reply by ChristopherPisz** [reply]
Went to Clion's File->settings->Build, Execution, Deployment->Python Interpreter and set that. Now it complains sometimes with the error above and sometimes about not being able to find fmtConfig,cmake.

I am guessing that I need to do the "using packages" step on Conan's site and that the generated project with its .py file is for "creating packages"

**Reply by friendlyanon** [reply]
cmake-init tells you exactly how to configure a project:

https://github.com/friendlyanon/cmake-init/blob/65af49607058a9d94e9f502d9a8bba393f37ae02/cmake-init/cmake_init.py#L334-L337

**Comment by ChristopherPisz** [comment]
Only thing I found in there was to execute `conan install . -s build_type=Debug -b missing`, from the project root director, which I did. I still end up with errors.

Here is a complete list of steps so far:
```
pip3 install cmake-init
sudo apt install clang-tidy
sudo apt install cppcheck
sudo apt install doxygen
sudo apt install lcov
sudo apt install clang-format-12
pip3 install codespell
pip3 install conan

cmake-init -p conan -s ./testproject
```



CLion
Open the folder in a new window
Open Project Wizard dialog should pop up
Change generator to “unix makefiles”
Click “dev preset” in the left pane
Check “enable profile”
Click OK

File->settings->Build, Execution, Deployment->Python Interpreter
Choose your Python3 interpreter in the drop down box

Conan Profile
```
conan profile new default --detect  # Generates default profile detecting GCC and sets old ABI
conan profile update settings.compiler.libcxx=libstdc++11 default # Sets libcxx to C++11 ABI
```

Get Dependencies
The generated library project we’ve made is currently dependent on the fmt library
We can see that from the conanfile.py which was generated as part of the create package requirements for the testproject.

From the root of the project directory, run the command
```conan install . -s build_type=Debug -b missing```

This will retrieve the missing packages for out dependencies into the local cache.

Result after reloading root's CmakeLists.txt:
```
/snap/clion/193/bin/cmake/linux/bin/cmake -DCMAKE_BUILD_TYPE=Debug -G "Unix Makefiles" -S /home/cpisz/git-workspace/testproject -B /home/cpisz/git-workspace/testproject/cmake-build-debug
CMake Error at CMakeLists.txt:60 (find_package):
  By not providing "Findfmt.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "fmt", but
  CMake did not find one.

  Could not find a package configuration file provided by "fmt" with any of
  the following names:

    fmtConfig.cmake
    fmt-config.cmake

  Add the installation prefix of "fmt" to CMAKE_PREFIX_PATH or set "fmt_DIR"
  to a directory containing one of the above files.  If "fmt" provides a
  separate development package or SDK, be sure it has been installed.
```

If I try to run any of  
```
cmake --preset=dev
cmake --build --preset=dev
ctest --preset=dev
```

from the Hacking.md, as it says, from the project root. I end up with 

```
CMake Error: The source directory "/home/cpisz/git-workspace/testproject/--preset=dev" does not exist.
```

If I try from proect root
```
cmake --preset=dev .
CMake Error at cmake/prelude.cmake:4 (message):
  In-source builds are not supported.  Please read the BUILDING document
  before trying to build this project.  You may need to delete
  'CMakeCache.txt' and 'CMakeFiles/' first.
Call Stack (most recent call first):
  CMakeLists.txt:3 (include)
```

If I create and cd to a build directory and
```
cmake --preset=dev ..
-- The CXX compiler identification is GNU 9.4.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
CMake Error at CMakeLists.txt:60 (find_package):
  By not providing "Findfmt.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "fmt", but
  CMake did not find one.

  Could not find a package configuration file provided by "fmt" with any of
  the following names:

    fmtConfig.cmake
    fmt-config.cmake
```

Sorry if I appear dumb. I have a 142 IQ, I promise. I've never used these tools before and am noobish to cmake and unix in general.

**Reply by ChristopherPisz** [reply]
I see there is a `fmt-config.cmake` in the testproject/conan directory, but evidently cmake doesn't know how to find it using the above steps alone.

**Reply by friendlyanon** [reply]
Best I can do is show you my process of opening a project using CLion:

* Generate project using `cmake-init -p conan -s proj`
* Install dependencies using `conan install . -s build_type=Debug -b missing` (from HACKING.md)
* Open folder using CLion
* Disable `Release` and `Debug` built-in profiles and select the `dev` preset only

This configures and builds fine for me.

**Reply by ChristopherPisz** [reply]
Went through your steps and that works.
The key difference is "disable Release and Debug built in profiles"

Thanks for your patience and your help!

Now I'll have to figure out how to make a conan server somewhere so company secrets are kept in the company when I upload a library. I'll have another thread for sticking points there if I run into some. Excited to use this. I really like the static analysis tools being built into the process.
